### PR TITLE
ci: Add CI/CD pipelines: linter + build

### PR DESCRIPTION
**Pull request by Toktar**

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+name: "Build"
+on:
```
Comment (ashcherbakov) [inline .github/workflows/build.yml:2]: When do we call these jobs? On every PR?

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      - edited
+      - synchronize
+      - ready_for_review
+      - review_requested
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  lint-pr:
```
Comment (ashcherbakov) [inline .github/workflows/pull-request.yml:20]: Is it the only job which's called on the PR? What about build and tests?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+          cache: true
+
+      - name: Run GoReleaser
+        uses: goreleaser/goreleaser-action@v3
+        with:
+          distribution: goreleaser
+          version: latest
+          args: build --rm-dist --snapshot --id ubuntu-latest-amd64
+
+      - name: Store artifact
```
Comment (ashcherbakov) [inline .github/workflows/build.yml:30]: If it's called for a PR, why do we need to store the artifacts?

File: .goreleaser.yml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      - "linux_amd64"
+    env:
+      - CGO_ENABLED=1
+    flags:
+      - -mod=readonly
+      - -tags="netgo ledger"
+      - -trimpath
+    ldflags:
+      - -s -w
+      - -X github.com/cosmos/cosmos-sdk/version.Name=canow-chain
```
Comment (ashcherbakov) [inline .goreleaser.yml:28]: I think this all should be part of the makefile, and the CI can be pretty trivial: `make build`.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+###############################################################################
+###                             Build / Install                             ###
+###############################################################################
+
+all: build
+
+install: go.sum go-version
+	go install -mod=readonly $(BUILD_FLAGS) $(CANOW_DIR)
+
+build: go.sum go-version
```
Comment (ashcherbakov) [inline Makefile:133]: What about unit tests?

File: docker/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
+
+# Inits node configuration and runs the node.
+# e -> exit immediately, u -> treat unset variables as errors and immediately, o -> sets the exit code to the rightmost command 
+set -euo pipefail
+
+# within the container, $HOME=/home/cheqd
+CHEQD_ROOT_DIR="$HOME/.cheqdnode"
```
Comment (ashcherbakov) [inline docker/entrypoint.sh:8]: Why cheqd? Should it be `CANOW_CHAIN_ROOT_DIR` with a value `.canowchain`?

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -1,53 +1,129 @@
...
+          flavor: |
+            latest=auto
+          tags: |
+            type=semver,pattern={{version}},value=${{ needs.release-binary.outputs.RELEASE_VERSION }}
+            type=raw,value=production-latest
+            type=sha,format=long
+          labels: |
+            org.opencontainers.image.vendor="Canow"
+            org.opencontainers.image.created={{date 'dddd, MMMM Do YYYY, h:mm:ss a'}}
+#            org.opencontainers.image.documentation="https://docs.cheqd.io/node"
```
Comment (ashcherbakov) [inline .github/workflows/release.yml:117]: Please get rid of `cheqd`

**Review by ashcherbakov — COMMENTED** [review]
Thanks for the great progress.
In general it looks good, please see my comments.
Also a couple more generic things:
1) It seems the PR covers lots of tickets and functionality. It would be much better to split it into multiple PRs: CI Build, CI linter, CI tests, CI release, Docker, integration tests, etc.
2) CI seems a bit overcomplicated to me. If we have `make` for build, tests, lint, why CI can't just call these simple commands? Also it's not clear what actions are done for a PR, what actions are done later (some naming are a bit confusing).

File: .github/workflows/pull-request.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+name: "PR Title Check"
+on:
+  pull_request_target:
+    branches:
+      - main
+      - develop
+      - release/**
+    types:
```
Comment (Toktar) [inline .github/workflows/pull-request.yml:8]: Simplify

File: .github/workflows/dispatch.yml [hunk 9]
```diff
@@ -0,0 +1,32 @@
+name: "Workflow Dispatch"
+on: push
```
Comment (Toktar) [inline .github/workflows/dispatch.yml:2]: Change to PR trigger

File: .github/workflows/dispatch.yml [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  call-test:
+    name: "Tests"
+    needs: call-build
+    uses: ./.github/workflows/test.yml
+    secrets: inherit
+
+  call-release:
+    name: "Release"
+    needs: call-test
+    if: ${{ github.ref_protected == true }}
```
Comment (Toktar) [inline .github/workflows/dispatch.yml:30]: change to tag creation

File: .github/workflows/pull-request.yml (hunk 8 above)
Comment (Toktar) [reply .github/workflows/pull-request.yml:8]: Looks like there are a lot of activity types and we don't want to trigger pipelines for all of them https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

File: .github/workflows/dispatch.yml (hunk 10 above)
Comment (Toktar) [reply .github/workflows/dispatch.yml]: Done

File: .github/workflows/dispatch.yml (hunk 9 above)
Comment (Toktar) [reply .github/workflows/dispatch.yml]: Done

File: .github/workflows/pull-request.yml (hunk 8 above)
Comment (Toktar) [reply .github/workflows/pull-request.yml]: Done

File: docker/localnet/gen-network-config.sh [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+# sed in macos requires extra argument
+if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+    SED_EXT=''
+elif [[ "$OSTYPE" == "darwin"* ]]; then
+    SED_EXT='.orig'
+fi
+
+# Params
+CHAIN_ID=${1:-"cheqd"} # First parameter, default is "cheqd"
```
[the remark this answers is not in the document]
Comment (Toktar) [reply docker/localnet/gen-network-config.sh]: Done

File: docker/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+FROM alpine:3.16 AS runner
+
+# Install pre-requisites
+RUN apk update && apk add --no-cache bash ca-certificates
+
+# Copy over binaries from the build-env
+COPY --from=builder /go/src/github.com/canow-co/canow-chain/build/canow-chain /bin/canow-chain
+
+# Set user directory and details
+ARG HOME_DIR="/home/cheqd"
```
[the remark this answers is not in the document]
Comment (Toktar) [reply docker/Dockerfile]: Done. Changed to `canow`

File: docker/localnet/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+    healthcheck:
+      test: curl -s http://localhost:26657/status | grep -c false
+      interval: 10s
+      timeout: 2s
+      retries: 6
+      start_period: 15s
+    volumes:
+      - type: volume
+        source: validator-0-data
+        target: /home/cheqd
```
[the remark this answers is not in the document]
Comment (Toktar) [reply docker/localnet/docker-compose.yml]: Done. Changed to `canow`

File: docker/entrypoint.sh (hunk 6 above)
Comment (Toktar) [reply docker/entrypoint.sh]: Fixed

File: Makefile (hunk 5 above)
Comment (Toktar) [reply Makefile:133]: `make test` and `make integration-test` will be added in scope of the next PR: https://github.com/canow-co/canow-chain/pull/5

**Review by ashcherbakov — APPROVED** [review]

**Pull request merged**